Recommend button query in concept search incompatible with SQL Server CDM

I posted this in the Atlas repo, but later thought it was probably more appropriate to post here. It is a 2.12 issue.

Expected behavior:
UI should return recommended concepts

Actual behavior

Error in log when pressing recommend. The problem is the LENGTH function, which is not valid SQL Server syntax--should be LEN:

broadsea-webtools2 | 2022-11-02 13:52:56.041 ERROR https-openssl-nio-8443-exec-4 org.ohdsi.webapi.util.GenericExceptionMapper - [] - org.springframework.jdbc.UncategorizedSQLException: PreparedStatementCallback; uncategorized SQLException for SQL [select top 100 CONCEPT_ID, CONCEPT_NAME, ISNULL(STANDARD_CONCEPT,'N') STANDARD_CONCEPT, ISNULL(INVALID_REASON,'V') INVALID_REASON, CONCEPT_CODE, CONCEPT_CLASS_ID, DOMAIN_ID, VOCABULARY_ID, VALID_START_DATE, VALID_END_DATE,
broadsea-webtools2 | LENGTH(REPLACE(REPLACE(CONCEPT_NAME, ' ',''), '-', '')) as c_length,
broadsea-webtools2 | LENGTH(REPLACE(REPLACE(REPLACE(REPLACE(lower(concept_name), ?,''), ?,''),' ',''),'-','')) as r_length,
broadsea-webtools2 | 1.0 - (1.0 * LENGTH(REPLACE(REPLACE(REPLACE(REPLACE(lower(concept_name), ?,''), ?,''),' ',''),'-','')) / LENGTH(REPLACE(REPLACE(CONCEPT_NAME, ' ',''), '-', ''))) as ratio_score
broadsea-webtools2 | from (
broadsea-webtools2 | select c1.concept_id as matched_concept
broadsea-webtools2 | from omop.concept c1
broadsea-webtools2 | where lower(concept_name) like ?
broadsea-webtools2 | union
broadsea-webtools2 | select cs1.concept_id as matched_concept
broadsea-webtools2 | from omop.concept_synonym cs1
broadsea-webtools2 | where lower(concept_synonym_name) like ?
broadsea-webtools2 | ) t1
broadsea-webtools2 | inner join omop.concept c1 on t1.matched_concept = c1.concept_id
broadsea-webtools2 | WHERE c1.standard_concept = 'S' AND (DOMAIN_ID IN (?,?,?) OR (DOMAIN_ID = 'Measurement' and LOWER(concept_class_id) in ('lab test', 'procedure')))
broadsea-webtools2 | order by ratio_score desc]; SQL state [S0010]; error code [195]; 'LENGTH' is not a recognized built-in function name.; nested exception is com.microsoft.sqlserver.jdbc.SQLServerException: 'LENGTH' is not a recognized built-in function name.
broadsea-webtools2 | at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:90)
broadsea-webtools2 | at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:82)
broadsea-webtools2 | at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:82)
broadsea-webtools2 | at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:654)
broadsea-webtools2 | at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:688)
broadsea-webtools2 | at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:720)
broadsea-webtools2 | at org.springframework.jdbc.core.JdbcTemplate.query(JdbcTemplate.java:770)
broadsea-webtools2 | at org.ohdsi.webapi.service.VocabularyService.executeSearch(VocabularyService.java:455)
broadsea-webtools2 | at sun.reflect.GeneratedMethodAccessor336.invoke(Unknown Source)
broadsea-webtools2 | at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
broadsea-webtools2 | at java.lang.reflect.Method.invoke(Method.java:498)
broadsea-webtools2 | at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81)
broadsea-webtools2 | at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
broadsea-webtools2 | at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
broadsea-webtools2 | at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
broadsea-webtools2 | at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
broadsea-webtools2 | at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
broadsea-webtools2 | at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
broadsea-webtools2 | at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
broadsea-webtools2 | at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
broadsea-webtools2 | at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
broadsea-webtools2 | at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
broadsea-webtools2 | at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
broadsea-webtools2 | at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
broadsea-webtools2 | at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
broadsea-webtools2 | at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
broadsea-webtools2 | at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
broadsea-webtools2 | at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
broadsea-webtools2 | at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473)
broadsea-webtools2 | at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427)
broadsea-webtools2 | at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388)
broadsea-webtools2 | at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341)
broadsea-webtools2 | at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:61)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
broadsea-webtools2 | at org.ohdsi.webapi.shiro.filters.ResponseNoCacheFilter.doFilter(ResponseNoCacheFilter.java:26)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.shiro.web.servlet.ProxiedFilterChain.doFilter(ProxiedFilterChain.java:66)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AbstractShiroFilter.executeChain(AbstractShiroFilter.java:450)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AbstractShiroFilter$1.call(AbstractShiroFilter.java:365)
broadsea-webtools2 | at org.apache.shiro.subject.support.SubjectCallable.doCall(SubjectCallable.java:90)
broadsea-webtools2 | at org.apache.shiro.subject.support.SubjectCallable.call(SubjectCallable.java:83)
broadsea-webtools2 | at org.apache.shiro.subject.support.DelegatingSubject.execute(DelegatingSubject.java:387)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AbstractShiroFilter.doFilterInternal(AbstractShiroFilter.java:362)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.executeChain(AdviceFilter.java:108)
broadsea-webtools2 | at org.apache.shiro.web.servlet.AdviceFilter.doFilterInternal(AdviceFilter.java:137)
broadsea-webtools2 | at org.apache.shiro.web.servlet.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:125)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.ohdsi.webapi.shiro.filters.CacheFilter.doFilter(CacheFilter.java:35)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.ohdsi.webapi.shiro.filters.ExceptionHandlerFilter.doFilter(ExceptionHandlerFilter.java:28)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
broadsea-webtools2 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:109)
broadsea-webtools2 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
broadsea-webtools2 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197)
broadsea-webtools2 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:128)
broadsea-webtools2 | at org.springframework.boot.web.support.ErrorPageFilter.access$000(ErrorPageFilter.java:66)
broadsea-webtools2 | at org.springframework.boot.web.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:103)
broadsea-webtools2 | at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
broadsea-webtools2 | at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:121)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
broadsea-webtools2 | at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
broadsea-webtools2 | at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
broadsea-webtools2 | at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
broadsea-webtools2 | at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
broadsea-webtools2 | at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
broadsea-webtools2 | at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
broadsea-webtools2 | at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)
broadsea-webtools2 | at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
broadsea-webtools2 | at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
broadsea-webtools2 | at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
broadsea-webtools2 | at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
broadsea-webtools2 | at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:890)
broadsea-webtools2 | at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1789)
broadsea-webtools2 | at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
broadsea-webtools2 | at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
broadsea-webtools2 | at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
broadsea-webtools2 | at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
broadsea-webtools2 | at java.lang.Thread.run(Thread.java:748)
broadsea-webtools2 | Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: 'LENGTH' is not a recognized built-in function name.
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:262)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1632)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.doExecutePreparedStatement(SQLServerPreparedStatement.java:602)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement$PrepStmtExecCmd.doExecute(SQLServerPreparedStatement.java:524)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:7418)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:3272)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:247)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:222)
broadsea-webtools2 | at com.microsoft.sqlserver.jdbc.SQLServerPreparedStatement.executeQuery(SQLServerPreparedStatement.java:446)
broadsea-webtools2 | at org.springframework.jdbc.core.JdbcTemplate$1.doInPreparedStatement(JdbcTemplate.java:696)
broadsea-webtools2 | at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:638)
broadsea-webtools2 | ... 121 more
Steps to reproduce behavior

Atlas and WebAPI 2.12.0

Click search on left frame.
Type a concept.
Search.
Select some results.
Press Recommend (shopping cart button).

![Screen Shot 2022-11-02 at 16 20 41](https://user-images.githubusercontent.com/4108940/201262968-c3fad4a7-22d4-4144-a609-b53d31a74d29.png)

--- a/src/main/resources/resources/vocabulary/sql/searchLexical.sql
+++ b/src/main/resources/resources/vocabulary/sql/searchLexical.sql
@@ -1,7 +1,7 @@
 select top 100 CONCEPT_ID, CONCEPT_NAME, ISNULL(STANDARD_CONCEPT,'N') STANDARD_CONCEPT, ISNULL(INVALID_REASON,'V') INVALID_REASON, CONCEPT_CODE, CONCEPT_CLASS_ID, DOMAIN_ID, VOCABULARY_ID, VALID_START_DATE, VALID_END_DATE,
-LENGTH(REPLACE(REPLACE(CONCEPT_NAME, ' ',''), '-', '')) as c_length,
-LENGTH(REPLACE(REPLACE(@replace_expression,' ',''),'-','')) as r_length,
-1.0 - (1.0 * LENGTH(REPLACE(REPLACE(@replace_expression,' ',''),'-','')) / LENGTH(REPLACE(REPLACE(CONCEPT_NAME, ' ',''), '-', ''))) as ratio_score
+LEN(REPLACE(REPLACE(CONCEPT_NAME, ' ',''), '-', '')) as c_length,
+LEN(REPLACE(REPLACE(@replace_expression,' ',''),'-','')) as r_length,
+1.0 - (1.0 * LEN(REPLACE(REPLACE(@replace_expression,' ',''),'-','')) / LEN(REPLACE(REPLACE(CONCEPT_NAME, ' ',''), '-', ''))) as ratio_score
 from (
   select c1.concept_id as matched_concept
   from @CDM_schema.concept c1

